messages/device-added-notification.json

```diff
@@ -38,7 +38,419 @@
+                      "title": "Property read-only",
+                      "description": "If the property is read-only"
+                    },
+                    "visible": {
+                      "type": "boolean",
+                      "title": "Property visible",
+                      "description": "If the property is visible"
+                    }
+                  }
+                }
+              }
+            },
+            "actions": {
+              "type": "object",
+              "patternProperties": {
+                ".*": {
+                  "type": "object",
+                  "title": "Action",
+                  "description": "Description of the Action",
+                  "properties": {
+                    "title": {
+                      "type": "string",
+                      "title": "Action title",
+                      "description": "The title of the Action"
+                    },
+                    "description": {
+                      "type": "string",
+                      "title": "Action Description",
+                      "description": "Description of the Action"
+                    },
+                    "links": {
+                      "type": "array",
+                      "items": {
+                        "type": "object",
+                        "title": "Action Link",
+                        "description": "An action link",
+                        "required": [
+                          "href",
+                          "rel"
+                        ],
+                        "properties": {
+                          "href": {
+                            "type": "string",
+                            "title": "Link href",
+                            "description": "The href of the link"
+                          },
+                          "rel": {
+                            "type": "string",
+                            "title": "Link rel",
+                            "description": "The type of the relationship"
+                          },
+                          "mediaType": {
+                            "type": "string",
+                            "title": "Link media type",
+                            "description": "The media type of the link"
+                          }
+                        }
+                      }
+                    },
+                    "input": {
+                      "type": "object",
+                      "title": "Action input",
+                      "description": "An action input",
+                      "required": [
+                        "type"
+                      ],
+                      "properties": {
+                        "type": {
+                          "type": "string",
+                          "title": "Input type",
+                          "description": "The type of the input",
+                          "enum": [
+                            "object"
```

Comment:
The action input can actually be anything.

Reply:
According to the [spec](https://iot.mozilla.org/wot/#action-object) the input is _An input object_.
The `object` here is the object which contains the metadata about the input.
If that would be a boolean what would `true` or `false` say about the metadata of the input?

Reply:
That mix of schema and meta schema is a bit confusing :smile:.
I looked [here](https://iot.mozilla.org/wot/#action-object) for the definition.
It looks a bit like the list contradicts the example.
The listing mentions that `An input object` with properties like `type`, `unit`, `minimum, and so on but the example shows that those properties are nested inside a properties object.
If I would just know the listing my example would like this:
```
"fade": {
  "@type": "FadeAction",
  "title": "Fade",
  "description": "Fade the lamp to a given level",
  "input": {
    "type": "object",
      "level": {
        "type": "integer",
        "minimum": 0,
        "maximum": 100
      },
      "duration": {
        "type": "integer",
        "minimum": 0,
        "unit": "milliseconds"
      }
  },
  "links": [{"href": "/things/lamp/actions/fade"}]
}
```

Reply:
Let's look at a full example
```
{
  "id": "42",
  "title": "The answer to life, the universe and everything",
  "actions": {
    "answer": {
      "title": "Answer",
      "input": {
          "type": "object",
          "properties": {
            "code": {
                "type": "integer"
            },
          }
      },
      "links": [{"href": "/things/lamp/actions/fade"}]
    }
  }
}
```
We are talking about the first type.
How would the example look like if that type would be `boolean`?

Reply:
Was the first type actually intended?
The listing clearly talks about the inner object where the type describes the actual value, just like in properties.
@benfrancis maybe you can bring light into the dark :smile:

Reply:
Ok, I found the implementation in the gateway.
I try to solve it using `oneOf`.

Reply:
I added support for primitive types.
I omitted the `null` type and the `array` type because the gateway doesn't support them.